DAPT codegen data curation tutorial

jvamaraju opened the pull request:
## Description
<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
Added a tutorial to curate data for DAPT codegen use-case with Nemo Curator

## Usage
<!-- Potentially add a usage example below -->
```python
codegen_DAPT\llm_data_collection\notebooks\data_curation-nemo-curator_DAPT.ipynb
```
## Checklist
<!--
Note: All commits need to be signed and signed off. This can be done via `-sS` flags while commiting
`git commit -sS -m "...."
-->
- I am familiar with the [Contributing Guide](https://github.com/NVIDIA/NeMo-Curator/blob/main/CONTRIBUTING.md).
- New or Existing tests cover these changes.
- The documentation is up to date with these changes.

Maghoumi commented:
@jvamaraju Thank you for your contribution. Here's some initial feedback after spending some time going through the PR.

### High level comments

Considering that this is a tutorial, we need to have explicit entry points for each topic where we are attempting to teach users how to accomplish certain things, or show them the best practices. We can follow a structure roughly similar to what you walked us through in the weekly meetings. Consider these (hypothetical) steps. I think the tutorial greatly benefits from having a "flow" where users can experience each data curation step by going through the flow:

* **Step 1: Dependency Installation**
  - Before users do anything, they need to understand all the code dependencies they need to fulfill. This includes pip package installs, PDF cleaner installation, etc. We could either have a notebook for this or a README file. I think a notebook would be better due to its interactive nature.

* **Step 2: Data Acquisition**
  - Substep 1: Users need to download the data from the Google Drive share. Show them where the data is located, and what directory this needs to be under.
  - Substep 2: Users need to run some shell scripts to traverse the downloaded data and create the relevant meta data.
  - ... [The rest follows] ...

* **Step 3: Data Curation using NeMo Curator**
  - Substep 1: Install dependencies
  - Substep 2: .....

* **Step 4: Data Curation without NeMo Curator**
  - Substep 1: .....
  - Substep 2: .....

Specifically, I think the data acquisition step also deserves a dedicated notebook where users can walk through the sequence of operations they need to do before they can follow along with the rest of the tutorial. Also, if there are shell scripts that need execution, we can make cells for them inside the notebook using `%` operator ([see this example](https://stackoverflow.com/a/70398447/398316)). This way, users won't have to worry about the order of executing different scripts.

Lastly, where relevant, I suggest clearing the cell outputs from the notebooks. I found myself scrolling a lot while going through the code and understanding the structure. For instance, [this notebook](https://github.com/jvamaraju/ndc_dapt_playbook/blob/main/tutorials/dapt-curation/codegen_DAPT/llm_data_collection/notebooks/data_curation-nemo-curator_DAPT.ipynb), is around 2MB and most cell outputs contain logs that users would be able to reproduce if they ran it locally. We should be safe to remove all those outputs from the checked in file. In some cases, it might make sense to keep that output (e.g. when we're trying to show a sample output to the user), so it's fine to keep them.

### Organization
The tutorial consists of 4 main types of files:
1. Python scripts
2. Shell scripts
3. Notebook files
4. Data files consisting of
   1. Images (e.g. PNG)
   2. JSONL (e.g. [this one](https://github.com/jvamaraju/ndc_dapt_playbook/blob/main/tutorials/dapt-curation/codegen_DAPT/llm_data_collection/notebooks/github_repos_opensource.jsonl))
   3. Data that the user downloads.

One challenge I faced while going through the repo was that different files appear in different places. In order to greatly enhance the readability and organization, I suggest the following structure (this is a rough sketch, feel free to incorporate based on your intuition of course!) :
* Aggregate all scripts inside one folder (e.g. `scripts`). It can have subfolders to organize based on the step of data curation (e.g. `scripts/1-data-acquisition` or `scripts/2-data-curation`).
* Organize Python scripts into logical units. For instance, if there are scripts that are helpers, they can all go inside a `helpers` directory. Reusable code can go into a `common` folder for instance.
* Aggregate all images in the same place. This has been already done, but some PNG files are still floating around.
* Create a centralized place for all other types of data. Perhaps the `dataset` folder that you have can serve as this? I'm thinking we could create a specific folder inside it `dataset/downloaded` where users can download the dataset under. Then everything else can reside in `dataset` directly (e.g. JSONL files).

### PDF Cleaner
This is a very useful tool and is standalone. I think it can be reused over and over throughout the repo. In its current form, it looks like it is a step of the tutorial. I'll defer to @ayushdg or others to comment on the best organization for this. It also contains docker image creation, etc. which appear to be out of the scope of the tutorial. Let's brainstorm on the most logical way of integrating this tool into the repo.

jvamaraju commented:
@Maghoumi  I have incorporated all the feedback. Specifically
Added license headers to all the files.
Cleaned up and unify function names
Added docstrings to all functions
Thoroughly tested with all the sources (currently it limits to 10 sources of Wikipedia, GitHub and arXiv).

Please review.

Maghoumi commented on tutorials/dapt-curation/README.md:
```diff
@@ -0,0 +1,67 @@
+This playbook utilizes specific tools and techniques. First, we convert all files to Txt format (if not already in Txt), compress files on disk, add metadata, and convert them to JSON (`./data/raw/`). Then, we leverage [NeMo Curator](https://github.com/NVIDIA/NeMo-Curator/tree/main) to mine high-quality text at scale from a massive code-generation corpus. [NeMo Curator](https://github.com/NVIDIA/ ...
+
+
+## NeMo Tools and Resources
```
Perhaps this is not needed? Because it's already inside the GitHub repo.

Maghoumi commented on tutorials/dapt-curation/README.md:
```diff
@@ -0,0 +1,67 @@
+* [NeMo-Curator GitHub repo](https://github.com/NVIDIA/NeMo-Curator/tree/main)
+
+
+## Software Requirements
```
The installation instructions are on the main README file already, so I suggest removing this as well to prevent the need for keeping it updated.

Maghoumi commented on tutorials/dapt-curation/README.md:
```diff
@@ -0,0 +1,67 @@
+
+
+## Software Requirements
+* This playbook has been tested on: nvcr.io/nvidia/nemo:24.03.01.framework. It is expected to work similarly on other environments. Within the container, from the NeMo-Curator directory install using `pip install --extra-index-url https://pypi.nvidia.com ".[cuda12x]"`
```
Let's add a note that this can be executed on a local machine with a GPU, but it has been tested with .....
Same as above, let's remove the install instructions.

Maghoumi commented on tutorials/dapt-curation/README.md:
```diff
@@ -0,0 +1,67 @@
+
+
+## Walkthrough
+For a detailed walkthrough of this tutorial, please see this blog post.
```
Let's remove this for now. I can add this back once the blog is published.

Maghoumi commented on tutorials/distributed_data_classification/distributed_data_classification.ipynb:
[For this and other files below that are irrelevant to the tutorial] I think the permission values changed? Could you please undo the changes to all these files and push again? Ideally, this PR shouldn't modify any irrelevant parts of the repo.

jvamaraju replied:
Can you check now? I did not change permissions I thought. Just synced the repo with main one again. Let me know if you still see this issue

Maghoumi replied:
These files are still showing as changed. Here's how I would resolve it:

```
git branch backup  # Make a backup of your branch as-is in case something gets messed up.
git reset 224800b4f141b41cdea376daf3c55a0cd788df47~1  # Undo your entire commit history, but keeps changed files.

git add tutorials/dapt-curation
git commit -sS -m "Your commit message"
git reset --hard  # to get rid of irrelevant changes
git push -f
```

Basically the idea is to only add the stuff under `tutorials/dapt-curation` to your branch, then commit everything as a new commit. Let me know if this makes sense.

jvamaraju replied:
Ok I did this. It seemed to have worked. Please let me know if you see it.

Maghoumi commented on tutorials/dapt-curation/README.md:
```diff
@@ -0,0 +1,67 @@
+
+* `\code\data` contains curated data after processing
+
+* Notebook for data curation with NeMo Curator: `\code\DAPT_data_curation.ipynb`
```
I'm a little confused about this: IIRC, the notebook version had issues, so we decided to publish this in the script format (`main.py`). If my understanding is correct, what's the reason for keeping the notebook?

If this is for content only, we can move the content to the blog.

jvamaraju replied:
@Maghoumi The notebook is just for content and quick tests. It will also help with any demos as we want to walk through the code. Should we still remove it?

Maghoumi replied:
Yes. I suggest keeping a copy of it in a branch or something for testing.
Given that this repo is production-grade, we're not allowed to push experimental/debugging content to it.

jvamaraju replied:
Ok I added it through .gitignore. Will this work?

Maghoumi commented on tutorials/dapt-curation/README.md:
```diff
@@ -0,0 +1,67 @@
+
+* Notebook for data curation with NeMo Curator: `\code\DAPT_data_curation.ipynb`
+
+The small size of this dataset makes it ideal for creating and validating data curation pipelines on a local machine or a cluster.
```
Is the dataset really small?

jvamaraju replied:
Yeah its extremely tiny compared to full dapt. We use only 100 wiki links, 10 papers and 6 repos. Chipnemo was trained on much more data.

Maghoumi replied:
Since the goal was to reproduce ChipNeMo's data curation, shouldn't we still include all the sources?
We can use the `limit=10` for the downloaders (along with a note for users) to maintain the ability of running the whole thing on a local machine.

jvamaraju replied:
The issue is we are legally not allowed to share all the sources in any format. So we are limited to this. This replicates data curation on a small, diversified dataset and is representative.

Maghoumi commented on tutorials/dapt-curation/code/sources/github_repos.jsonl:
```diff
@@ -0,0 +1,6 @@
+"google/skywater-pdk-libs-sky130_fd_sc_lp"
```
Didn't we have way more repositories in this file? Did we remove for simplicity, or was there some other reason?

Asking because if someone wants to repro ChipNeMo data exactly, they'd need all the repos that was used for model training.

jvamaraju replied:
@Maghoumi We cannot share the full data that ChipNeMo used because of legal permissions. So, we had to sample only a few datasets where permissions to share was easy to get.

Maghoumi replied:
I thought the legal issue was with bundling all the data into the Google Drive share. Are you saying we can't include a link to the repositories, Wikipedia pages and arXiv papers?

jvamaraju replied:
Yes exactly. Not all github repos and arxiv papers have permissions for us to share. So I had to change to a few sources where we have permissions.

Maghoumi commented on tutorials/dapt-curation/code/utils.py:
```diff
@@ -0,0 +1,389 @@
+                text_field="text",
+                score_type=float,
+            ),
+            #             # If the document shrinks by > 30% in terms of number of lines after
```
Either uncomment or remove entirely.

Maghoumi commented on tutorials/dapt-curation/code/utils.py:
```diff
@@ -0,0 +1,389 @@
+    return cleaners(dataset)
+
+
+def filter_txt(dataset: DocumentDataset) -> DocumentDataset:
```
[nit] rename to `filter_text_sources` so that it's more descriptive?

jvamaraju replied:
We have filter_code and filter_txt. I think that's descriptive enough?

Maghoumi replied:
Yes, but let's rename to `filter_text`.

jvamaraju replied:
done

Maghoumi commented on tutorials/dapt-curation/code/utils.py:
```diff
@@ -0,0 +1,389 @@
+    for doc in json_docs:
+        print(doc)
+        if "text" in doc:
+            # print("initial", doc["text"])
```
Remove

Maghoumi commented on tutorials/dapt-curation/code/utils.py:
```diff
@@ -0,0 +1,389 @@
+
+                doc["text"] = doc["text"].replace(comment, redacted_text_string)
+
+                # print("final", doc["text"])
```
Remove

Maghoumi commented on tutorials/dapt-curation/code/utils.py:
```diff
@@ -0,0 +1,389 @@
+    return filtered_dataset
+
+
+# +
```
Remove `#`

Maghoumi commented on tutorials/dapt-curation/code/utils.py:
```diff
@@ -0,0 +1,389 @@
+                    pd.DataFrame([{"text": comment}]), npartitions=1
+                )
+                # Create a DocumentDataset instance
+                comment_doc_dataset = DocumentDataset(comment_df)
```
I'm a little confused as why this is being done this way. Couldn't we do this as part of the `run_curation_pipeline`? My concern is that this will have poor performance.

jvamaraju replied:
It has poor performance. But also, redaction has a few bugs when used for code files. We discussed this with Arham as well. The idea is to redact only the comment lines and not all the code. We saw that proper names within code get redaxted/replaced otherwise. Is there an alternate way?

Maghoumi replied:
Adding @ryantwolf and @ayushdg to comment on how to workaround this issue.

ryantwolf replied:
As far as I know the PII tool was made to operate on normal text documents and not on code. In particular, it assumes it can operate on the entire document. So, there will probably need to be a workaround. However, I can suggest on how to improve performance. I would recommend you construct the entire `DocumentDataset` filled with every comment in your dataset. Then, perform the PII on the comments, then replace them. Doing this in a batched manner should allow the PII removal to perform better. @singhva if you have any additional thoughts

jvamaraju replied:
Thanks @ryantwolf  and @Maghoumi . I have updated the PII reaction piece now.

Maghoumi commented on tutorials/dapt-curation/code/utils.py:
```diff
@@ -0,0 +1,389 @@
+            supported_entities=[
+                "PERSON",
+                "EMAIL_ADDRESS",
+                #                                "URL",
```
Either uncomment, or remove entirely.

jvamaraju replied:
Removed it, but it shows the developer the other possible options as well

Maghoumi replied:
Thanks.
The usage is already documented and shown in other tutorials.

Maghoumi commented on tutorials/dapt-curation/code/utils.py:
```diff
@@ -0,0 +1,389 @@
+                #                                "LOCATION"
+            ],
+            anonymize_action="replace",
+            #             batch_size=1000,
```
Remove?

jvamaraju replied:
This is a good toggle option to keep. Again, points to hyperparameters users need to play with.

Maghoumi replied:
Then we should add `if .... else.... ` block. Leaving commented out code is not allowed in production grade code.

jvamaraju replied:
ok removed it. Its a tutorial right? But i understand.

Maghoumi commented on tutorials/dapt-curation/code/utils.py:
```diff
@@ -0,0 +1,389 @@
+    Discard text files based on lines.
+    """
+    dataset_df = dataset.df
+    #     print(len(dataset_df.index))
```
Remove?
Since there appears to be multiple occurances, I won't comment further on this, please remove all debugging code from the PR for cleaning purposes.

jvamaraju replied:
done

Maghoumi commented on tutorials/dapt-curation/code/main.py:
```diff
@@ -0,0 +1,245 @@
+DATA_DIR = os.path.join(SCRIPT_DIR_PATH, "data")
+
+#
+# TODO
```
Could you add a note about this in the README file?

jvamaraju replied:
done

Maghoumi replied:
Did you also remove the TODO notes here? If already removed, it's not showing up here for some reason.

Maghoumi commented on tutorials/dapt-curation/code/main.py:
```diff
@@ -0,0 +1,245 @@
+            filter_code_lines,
+            filter_code,
+            clean_and_unify,
+            #          redact_pii,
```
Remove?

jvamaraju replied:
Lets keep this for now. I'm testing it even more for performance.

jvamaraju replied:
Added it in.

Maghoumi commented on tutorials/dapt-curation/code/main.py:
```diff
@@ -0,0 +1,245 @@
+    orig_df = orig_dataset_text.df.compute()
+    orig_df = orig_df.reset_index()
+
+    # Create a histogram for different file types -text
```
Let's move the logic for showing histograms, etc to its own function so that the code looks cleaner.

Something like `def plot_data(...)`

jvamaraju replied:
done

Maghoumi reviewed (commented):
Thanks again for your contribution. I took one pass and left some comments.

jvamaraju commented:
@Maghoumi  Please look at my comments and updated repo.

jvamaraju commented:
[like]  Janaki Vamaraju reacted to your message:
________________________________
From: Mehran Maghoumi ***@***.***>
Sent: Thursday, July 11, 2024 6:15:41 PM
To: NVIDIA/NeMo-Curator ***@***.***>
Cc: Janaki Vamaraju ***@***.***>; Mention ***@***.***>
Subject: Re: [NVIDIA/NeMo-Curator] DAPT codegen data curation tutorial (PR #88)


@Maghoumi commented on this pull request.

________________________________

On tutorials/distributed_data_classification/distributed_data_classification.ipynb<https://github.com/NVIDIA/NeMo-Curator/pull/88#discussion_r1674457723>:

These files are still showing as changed. Here's how I would resolve it:

git branch backup  # Make a backup of your branch as-is in case something gets messed up.
git reset 224800b4f141b41cdea376daf3c55a0cd788df47~1  # Undo your entire commit history, but keeps changed files.

git add tutorials/dapt-curation
git commit -sS -m "Your commit message"
git reset --hard  # to get rid of irrelevant changes
git push -f


Basically the idea is to only add the stuff under tutorials/dapt-curation to your branch, then commit everything as a new commit. Let me know if this makes sense.

________________________________

In tutorials/dapt-curation/README.md<https://github.com/NVIDIA/NeMo-Curator/pull/88#discussion_r1674440723>:

> @@ -0,0 +1,67 @@
+# Data curation for DAPT (Domain Adaptive Pre-Training)
+
+[ChipNeMo](https://arxiv.org/pdf/2311.00176) is a chip design domain adapted LLM. LLama models are continually pre-trained with 20B plus tokens on domain-specific chip design data, including code, documents, etc., based on NeMo foundation models and then fine-tuned with instruction datasets from design data as well as external sources.
+
+Here, we share a tutotial with best practices on data curation for DAPT (domain-adaptive pre-training) for a ChipNeMo-like code generation use case.
+
+* In this tutorial, we will leverage chip domain/hardware datasets from open-source GitHub repositories (`.code/sources/github_repos.jsonl`), wiki URLs (`.code/sources/wikipedia_urls.jsonl`), and academic papers(`./sources/arxiv_urls.jsonl`).
+
+* `\code\data` contains curated data after processing
+
+* Notebook for data curation with NeMo Curator: `\code\DAPT_data_curation.ipynb`


Yes. I suggest keeping a copy of it in a branch or something for testing.
Given that this repo is production-grade, we're not allowed to push experimental/debugging content to it.

________________________________

In tutorials/dapt-curation/README.md<https://github.com/NVIDIA/NeMo-Curator/pull/88#discussion_r1674442399>:

> @@ -0,0 +1,67 @@
+# Data curation for DAPT (Domain Adaptive Pre-Training)
+
+[ChipNeMo](https://arxiv.org/pdf/2311.00176) is a chip design domain adapted LLM. LLama models are continually pre-trained with 20B plus tokens on domain-specific chip design data, including code, documents, etc., based on NeMo foundation models and then fine-tuned with instruction datasets from design data as well as external sources.
+
+Here, we share a tutotial with best practices on data curation for DAPT (domain-adaptive pre-training) for a ChipNeMo-like code generation use case.
+
+* In this tutorial, we will leverage chip domain/hardware datasets from open-source GitHub repositories (`.code/sources/github_repos.jsonl`), wiki URLs (`.code/sources/wikipedia_urls.jsonl`), and academic papers(`./sources/arxiv_urls.jsonl`).
+
+* `\code\data` contains curated data after processing
+
+* Notebook for data curation with NeMo Curator: `\code\DAPT_data_curation.ipynb`
+
+The small size of this dataset makes it ideal for creating and validating data curation pipelines on a local machine or a cluster.


Since the goal was to reproduce ChipNeMo's data curation, shouldn't we still include all the sources?
We can use the limit=10 for the downloaders (along with a note for users) to maintain the ability of running the whole thing on a local machine.

________________________________

In tutorials/dapt-curation/code/sources/github_repos.jsonl<https://github.com/NVIDIA/NeMo-Curator/pull/88#discussion_r1674446964>:

> @@ -0,0 +1,6 @@
+"google/skywater-pdk-libs-sky130_fd_sc_lp"


I thought the legal issue was with bundling all the data into the Google Drive share. Are you saying we can't include a link to the repositories, Wikipedia pages and arXiv papers?

________________________________

In tutorials/dapt-curation/code/utils.py<https://github.com/NVIDIA/NeMo-Curator/pull/88#discussion_r1674448258>:

> +
+    Returns:
+        DocumentDataset: The cleaned and unified dataset.
+    """
+    cleaners = Sequential(
+        [
+            # Unify all the quotation marks
+            Modify(QuotationUnifier(), text_field="text"),
+            # Unify all unicode
+            Modify(UnicodeReformatter(), text_field="text"),
+        ]
+    )
+    return cleaners(dataset)
+
+
+def filter_txt(dataset: DocumentDataset) -> DocumentDataset:


Yes, but let's rename to filter_text.

________________________________

In tutorials/dapt-curation/code/utils.py<https://github.com/NVIDIA/NeMo-Curator/pull/88#discussion_r1674449907>:

> +        print(doc)
+        if "text" in doc:
+            # print("initial", doc["text"])
+
+            # Extract the comment content between /* and */
+            start_idx = doc["text"].find("/*")
+            end_idx = doc["text"].find("*/")
+            if start_idx != -1 and end_idx != -1:
+                comment = doc["text"][start_idx : end_idx + 2]
+
+                # Create a Dask DataFrame from the given data
+                comment_df = dd.from_pandas(
+                    pd.DataFrame([{"text": comment}]), npartitions=1
+                )
+                # Create a DocumentDataset instance
+                comment_doc_dataset = DocumentDataset(comment_df)


Adding @ryantwolf<https://github.com/ryantwolf> and @ayushdg<https://github.com/ayushdg> to comment on how to workaround this issue.

________________________________

In tutorials/dapt-curation/code/utils.py<https://github.com/NVIDIA/NeMo-Curator/pull/88#discussion_r1674459060>:

> +    """
+    Redacts personally identifiable information (PII) from a given dataset.
+
+    Args:
+        dataset (DocumentDataset): The dataset containing documents with PII.
+
+    Returns:
+        DocumentDataset: The redacted dataset with PII replaced by a generic value.
+    """
+    DEFAULT_MAX_DOC_SIZE = 3000000
+    redactor = Modify(
+        PiiModifier(
+            supported_entities=[
+                "PERSON",
+                "EMAIL_ADDRESS",
+                #                                "URL",


Thanks.
The usage is already documented and shown in other tutorials.

________________________________

In tutorials/dapt-curation/code/utils.py<https://github.com/NVIDIA/NeMo-Curator/pull/88#discussion_r1674450814>:

> +    Returns:
+        DocumentDataset: The redacted dataset with PII replaced by a generic value.
+    """
+    DEFAULT_MAX_DOC_SIZE = 3000000
+    redactor = Modify(
+        PiiModifier(
+            supported_entities=[
+                "PERSON",
+                "EMAIL_ADDRESS",
+                #                                "URL",
+                #                                "PHONE_NUMBER",
+                #                                "ADDRESS",
+                #                                "LOCATION"
+            ],
+            anonymize_action="replace",
+            #             batch_size=1000,


Then we should add if .... else.... block. Leaving commented out code is not allowed in production grade code.

________________________________

In tutorials/dapt-curation/code/main.py<https://github.com/NVIDIA/NeMo-Curator/pull/88#discussion_r1674443239>:

> +from nemo_curator.modifiers.pii_modifier import PiiModifier
+from nemo_curator.modifiers.unicode_reformatter import UnicodeReformatter
+from nemo_curator.modules import ExactDuplicates
+from nemo_curator.modules.modify import Modify
+from nemo_curator.utils.distributed_utils import get_client
+from nemo_curator.utils.file_utils import (
+    get_all_files_paths_under,
+    separate_by_metadata,
+)
+from nemo_curator.utils.script_utils import add_distributed_args, parse_client_args
+
+SCRIPT_DIR_PATH = os.path.dirname(os.path.abspath(__file__))
+DATA_DIR = os.path.join(SCRIPT_DIR_PATH, "data")
+
+#
+# TODO


Did you also remove the TODO notes here? If already removed, it's not showing up here for some reason.

—
Reply to this email directly, view it on GitHub<https://github.com/NVIDIA/NeMo-Curator/pull/88#discussion_r1674457723>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BIZL5HQL7OOUALHWEXGZI5LZL3DU3AVCNFSM6AAAAABIOERV7GVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCNZSGYZDCMRYHE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

jvamaraju commented:
pre-commit.ci autofix

jvamaraju commented:
pre-commit.ci autofix

ryantwolf commented:
Closing in favor of #160
